relayer/relayer.go

```diff
@@ -116,34 +115,71 @@ func NewRelayer(
-		m := response.Message()
-		requestID, ok := message.GetRequestID(m)
-		if !ok {
-			response.OnFinishedHandling()
-			return
+	// Get the latest processed block height from the database.
+	latestProcessedBlockData, err := r.db.Get(r.sourceChainID, []byte(database.LatestProcessedBlockKey))
+
+	// The following cases are treated as successful:
+	// 1) The database contains the latest processed block data for the chain
+	//    - In this case, we parse the block height and process warp logs from that height to the current block
+	// 2) The database has been configured for the chain, but does not contain the latest processed block data
+	//    - In this case, we save the current block height in the database, but do not process any historical warp logs
+	if err == nil {
+		// If the database contains the latest processed block data, then back-process all warp messages from that block to the latest block
+		// Note that the retrieved latest processed block may have already been partially (or fully) processed by the relayer on a previous run. When
+		// processing a warp message in real time, which is when we update the latest processed block in the database, we have no way of knowing
+		// if that is the last warp message in the block
+		latestProcessedBlock, success := new(big.Int).SetString(string(latestProcessedBlockData), 10)
+		if !success {
+			r.logger.Error("failed to convert latest block to big.Int", zap.Error(err))
+			return nil, nil, err
+		}
+
+		err = sub.ProcessFromHeight(latestProcessedBlock)
+		if err != nil {
+			logger.Warn(
+				"Encountered an error when processing historical blocks. Continuing to normal relaying operation.",
+				zap.String("chainID", r.sourceChainID.String()),
+				zap.Error(err),
+			)
 		}
-		r.messageChanLock.RLock()
-		messageResponseChan, ok := r.messageChanMap[requestID]
-		r.messageChanLock.RUnlock()
-		if ok {
-			messageResponseChan <- response
-		} else {
-			response.OnFinishedHandling()
+		return &r, sub, nil
+	}
+	if errors.Is(err, database.ErrChainNotFound) || errors.Is(err, database.ErrKeyNotFound) {
+		// Otherwise, latestProcessedBlock is nil, so we instead store the latest block height.
+		logger.Info(
+			"Latest processed block not found in database. Starting from latest block.",
+			zap.String("chainID", r.sourceChainID.String()),
+		)
+
+		err := sub.SetProcessedBlockHeightToLatest()
+		if err != nil {
+			logger.Warn(
+				"Failed to update latest processed block. Continuing to normal relaying operation",
+				zap.String("chainID", r.sourceChainID.String()),
+				zap.Error(err),
+			)
 		}
+		return &r, sub, nil
 	}
+
+	// If neither of the above conditions are met, then we return an error
+	r.logger.Warn(
+		"failed to get latest block from database",
+		zap.String("chainID", r.sourceChainID.String()),
+		zap.Error(err),
+	)
+	return nil, nil, err
 }
 
 // RelayMessage relays a single warp message to the destination chain. Warp message relay requests are concurrent with each other,
```

Comment:
"Warp message relay requests are concurrent with each other"

Does this need to be updated after the recent change to process messages from a given source serially?

Reply:
Nice catch - updated now